### PR TITLE
cudaPackages.cudnn: hash correction and 8.9.6.50 -> 8.9.7.29

### PR DESCRIPTION
**Pull request by ConnorBaker**
## Description of changes

- Fixes bad hash for release for `x86_64-linux` CUDA 10.2, v7.6.5.32 caused by bad copy-paste
- Updates 8.9.6.50 -> 8.9.7.29

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by SomeoneSerge** [comment]
I don't think `fq7IA5osMKsLx1jTA1iHZ2k972v0myJIWiwAvy4TbLN` had anything to do with bad copy-pasting btw. That's the same hash I saw in the git blames, so either we broke it much earlier, or nvidia really has swapped the blob. Otherwise, we really should avoid editing these files manually. Instead, we want a deterministic and idempotent update script that we can run pre-commit style, instead of reviews. If two users run the updater and get the exact same output, the hashes pass the review. We should also add this to the CI probably

**Comment by ConnorBaker** [comment]
> [!Note]
>
> Template `nixpkgs-review` command:
>
> ```console
> PR=273604; \
> SYSTEM="aarch64-linux"; \
> CUDA_SUPPORT="true"; \
> CUDA_CAPABILITIES='[ "7.5" ]'; \
> nixpkgs-review pr "$PR" \
>   --system "$SYSTEM" \
>   --no-shell \
>   --checkout commit \
>   --allow aliases \
>   --build-args "--max-jobs 1" \
>   --extra-nixpkgs-config "{
>     allowUnfree = true;
>     allowBroken = false;
>     cudaSupport = ${CUDA_SUPPORT:-false};
>     cudaCapabilities = ${CUDA_CAPABILITIES:-[]};
>   }"
> ```

## `aarch64-darwin`

In progress.

## `x86_64-darwin`

In progress.

## `aarch64-linux`

### Jetson

In progress.

### Non-Jetson

In progress.

### Non-CUDA

In progress.

## `x86_64-linux`

### Non-Jetson

In progress.

### Non-CUDA

In progress.

**Comment by ConnorBaker** [comment]
> I don't think `fq7IA5osMKsLx1jTA1iHZ2k972v0myJIWiwAvy4TbLN` had anything to do with bad copy-pasting btw. That's the same hash I saw in the git blames, so either we broke it much earlier, or nvidia really has swapped the blob. Otherwise, we really should avoid editing these files manually. Instead, we want a deterministic and idempotent update script that we can run pre-commit style, instead of reviews. If two users run the updater and get the exact same output, the hashes pass the review. We should also add this to the CI probably

I agree, but I don't have the bandwidth at the moment to work on that.

I imagine such a script could be implemented as a DFS traversal of the pages at the CUDNN resist URL. If there's a new version which doesn't exist, add it. If there's a newer version of the same major/minor release, replace it with the newer version. Otherwise, skip it.

**Comment by ConnorBaker** [comment]
@SomeoneSerge can you confirm that jaxlib and cupy are broken on master?

**Comment by SomeoneSerge** [comment]
@ConnorBaker yes https://hercules-ci.com/github/SomeoneSerge/nixpkgs-cuda-ci/jobs/7448

**Review by jonringer — COMMENTED** [review]
otherwise seems fine

File: pkgs/development/cuda-modules/cudnn/releases.nix [hunk 1]
```diff
@@ -244,18 +244,18 @@
...
       }
       {
-        version = "8.9.6.50";
+        version = "8.9.7.29";
         minCudaVersion = "12.0";
         maxCudaVersion = "12.2";
-        url = "https://developer.download.nvidia.com/compute/cudnn/redist/cudnn/linux-x86_64/cudnn-linux-x86_64-8.9.6.50_cuda12-archive.tar.xz";
-        hash = "sha256-FyIlnblSZbs4E0OKWhxuzZed6JrkU2YDkEBC4STTAtU=";
+        url = "https://developer.download.nvidia.com/compute/cudnn/redist/cudnn/linux-x86_64/cudnn-linux-x86_64-8.9.7.29_cuda12-archive.tar.xz";
+        hash = "sha256-R1MzYlx+QqevPKCy91BqEG4wyTsaoAgc2cE++24h47s=";
```
[the remark this answers is not in the document]
Comment (ConnorBaker) [reply pkgs/development/cuda-modules/cudnn/releases.nix:258]: @jonringer I don't understand how that could break 😟

I just re-ran `nix store prefetch-file` with the URL and it gave me the same hash back.

```console
$ nix store prefetch-file https://developer.download.nvidia.com/compute/cudnn/redist/cudnn/linux-x86_64/cudnn-linux-x86_64-8.9.7.29_cuda12-archive.tar.xz
Downloaded 'https://developer.download.nvidia.com/compute/cudnn/redist/cudnn/linux-x86_64/cudnn-linux-x86_64-8.9.7.29_cuda12-archive.tar.xz' to '/nix/store/2m4cq2z7dv41bylagmkq47lbybqfzy61-cudnn-linux-x86_64-8.9.7.29_cuda12-archive.tar.xz' (hash 'sha256-R1MzYlx+QqevPKCy91BqEG4wyTsaoAgc2cE++24h47s=').
```

[the remark this answers is not in the document]
Comment (jonringer) [reply pkgs/development/cuda-modules/cudnn/releases.nix:258]: must have been a transitory issue. I ran this again and it was able to download it fine.

**Review by jonringer — APPROVED** [review]

File: pkgs/development/cuda-modules/cudnn/releases.nix (hunk 1 above)
[the remark this answers is not in the document]
Comment (SomeoneSerge) [reply pkgs/development/cuda-modules/cudnn/releases.nix:258]: (now that's reassuring 😅 )

[the remark this answers is not in the document]
Comment (jonringer) [reply pkgs/development/cuda-modules/cudnn/releases.nix:258]: It's a pretty big download (800MB +).... who knows...

**Pull request merged**